File: Data Structures & Algorithms/two-integer-sum/submission-2.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        //brute force
        //o(n^2) and o(1)
        // for (int i = 0; i < nums.length; i++) {
        //     for (int j = i + 1; j < nums.length; j++) {
        //         if (nums[i] + nums[j] == target) {
        //             return new int[] {i,j};
        //         }
        //     }
        // }
        // return new int[] {};

        //Better solution
        //O(n) and O(n)
        HashMap<Integer, Integer> numToIndex = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            int diff = target - num;

            if (numToIndex.containsKey(diff)) {
                return new int[] {numToIndex.get(diff), i};
            }
            numToIndex.put(num, i);
        }
    return new int[]{};

    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row = 0
        while row < len(matrix):
            l, r = 0, len(matrix[row]) - 1
            while l <= r:
                if matrix[row][r] < target:
                    break
                else:
                    m = (l+r) // 2
                    if matrix[row][m] > target:
                        r = m - 1
                    elif matrix[row][m] < target:
                        l = m + 1
                    else:
                        return True
            row += 1
        return False